airbyte-cdk/python/airbyte_cdk/sources/connector_state_manager.py

```diff
@@ -14,41 +27,105 @@ class ConnectorStateManager:
     interface. It also provides methods to extract and update state
     """
 
-    # In the immediate, we only persist legacy which will be used during abstract_source.read(). In the subsequent PRs we will
-    # initialize the ConnectorStateManager according to the new per-stream interface received from the platform
-    def __init__(self, state: Union[List[AirbyteStateMessage], MutableMapping[str, Any]] = None):
-        if not state:
-            self.legacy = {}
-        elif self.is_migrated_legacy_state(state):
-            # The legacy state format received from the platform is parsed and stored as a single AirbyteStateMessage when reading
-            # the file. This is used for input backwards compatibility.
-            self.legacy = state[0].data
-        elif isinstance(state, MutableMapping):
-            # In the event that legacy state comes in as its original JSON object format, no changes to the input need to be made
-            self.legacy = state
-        else:
-            raise ValueError("Input state should come in the form of list of Airbyte state messages or a mapping of states")
+    def __init__(self, stream_instance_map: Mapping[str, Stream], state: Union[List[AirbyteStateMessage], MutableMapping[str, Any]] = None):
+        shared_state, per_stream_states = self._extract_from_state_message(state, stream_instance_map)
+        self.shared_state = shared_state
+        self.per_stream_states = per_stream_states
 
-    def get_stream_state(self, namespace: str, stream_name: str) -> AirbyteStateBlob:
-        # todo implement in upcoming PRs
-        pass
+    def get_stream_state(self, stream_name: str, namespace: Optional[str]) -> Mapping[str, Any]:
+        """
+        Retrieves the state of a given stream based on its descriptor (name + namespace) including any global shared state.
+        Stream state takes precedence over shared state when there are conflicts
+        :param stream_name: Name of the stream being fetched
+        :param namespace: Namespace of the stream being fetched
+        :return: The combined shared state and per-stream state of a stream
+        """
+        combined_state = {}
+        if self.shared_state:
```

Comment:
is it safe to assume that we can combine the global state with the per-stream state? I don't see something in the [V2 docs](https://docs.airbyte.com/understanding-airbyte/airbyte-protocol#v2-coming-soon) which indicates that it is. This appears to be merging keys from one namespace (global) to another (per-stream) which, I'm not sure that the behavior here is defined by the protocol?

Reply:
as per offline discussion, the acceptable path forward is going to still be to support global state coming in, however, if the `shared_state` value is `None`, we should throw an error since this is unintended behavior at this point in time. 

API connectors will not be using shared state and if that field is ever populated, something has gone wrong and we should stop syncing.